### DISCUSSION TITLE
refresh_token expiration

### DISCUSSION THREAD

**Discussion by olibanjoli** [timdorr/tesla-api#587] [General]
how long is the refresh_token valid? the expiration in the response is only for the access_token i guess. so is the refresh token forever valid? (except when changing tesla pw)

**Comment by snappyrain** [comment]
same question. anyone could answer ?

**Reply by olibanjoli** [reply]
i do not have a final answer but i just refreshed tokens using a refreshtoken which is 2 months old and bearer token was expired since 1 month.

so looks like there is plenty of time to use the refresh token

**Reply by briangru** [reply]
In the past, the access token expired in 45 days.  There was no information about the refresh token's lifetime, but it was probably at least that long.  I wouldn't assume anything beyond 45 days.  I've been assuming 8 hours, and if it works for a longer time, great.

**Comment by cch1** [comment]
BTTT - this information would be really helpful.  Even anecdotal evidence.  So far, the best info I have is @briangru 's comment.

**Comment by Matthew1471** [comment]
>The refresh token is single use only and expires after 3 months.

https://developer.tesla.com/docs/fleet-api/authentication/third-party-tokens

What that means in terms of number of seconds, whether it's calendar days or what is anyone's guess.
